Add file extensions to relative paths

lukaw3d opened the pull request:
Fixes https://github.com/oasisprotocol/docs/issues/4

Commit:
Use https links where possible

Commit:
Add markdown linter

Commit:
Check relative links have a file extension, and file exists

tjanez reviewed (commented):
Great work!

Could you also add Markdownlint check to CI? It would be cool if there is a Markdownlint GitHub action that would also annotate the errors, otherwise `yarn lintmarkdown` is OK.

Also, could you change "Auto-fix, mostly" commit message to something more concrete :slightly_smiling_face: .

Commit:
Use absolute paths for remapped links

Docusaurus config remaps some links to different routeBasePath, so linter can't
directly find referenced files.

lukaw3d commented:
Improved commit messages

tjanez reviewed (approved):
Thanks for tidying up the commits!

> Could you also add Markdownlint check to CI? It would be cool if there is a Markdownlint GitHub action that would also annotate the errors, otherwise `yarn lintmarkdown` is OK.

I've opened a follow up issue for that: #29.

Commit:
Manually fix relative links that appear twice in one line

lukaw3d commented on package.json:
```diff
@@ -12,7 +12,9 @@
-    "blc": "broken-link-checker --recursive --ordered --exclude https://t.me --exclude https://software.intel.com/content --exclude https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf --exclude https://support.ledger.com/hc/en-us/articles --exclude https://www.linkedin.com --exclude https://csrc.nist.gov/publications/detail/fips/180/4/final --exclude https://www.coinspeaker. ...
+    "blc": "broken-link-checker --recursive --ordered --exclude https://t.me --exclude https://software.intel.com/content --exclude https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf --exclude https://support.ledger.com/hc/en-us/articles --exclude https://www.linkedin.com --exclude https://csrc.nist.gov/publications/detail/fips/180/4/final --exclude https://www.coinspeaker. ...
+    "lint-markdown": "markdownlint-cli2 'docs/**/*.md'",
+    "fmt-markdown": "markdownlint-cli2-fix 'docs/**/*.md'"
```
oops, this missed links in `.mdx` files

Commit:
Auto-fix relative links (yarn fmt-markdown)

lukaw3d commented:
I also checked referenced external repos are okay, with some manual steps:
- remove .markdownlint.yml files in external repos
- add to .markdownlint-cli2.cjs
  ```js
    noInlineConfig: true,
    rules: {
        'emphasis-style': false,
        'code-block-style': false,
        'no-space-in-code': false,
        'single-title': false,
  ```
- run `yarn markdownlint-cli2-fix 'external/*/docs/**/*.md' 'external/*/docs/**/*.mdx'`

lukaw3d commented:
Running above steps + `yarn markdownlint-cli2 'external/**/*.md' 'external/**/*.mdx'` found these broken links:

external/oasis-core-ledger/CHANGELOG.md:90  [File missing] [Context: "doc/versioning.md"]
external/oasis-core-ledger/CHANGELOG.md:179  [File missing] [Context: "doc/versioning.md"]
external/oasis-core/CHANGELOG.md:1326  [File missing] [Context: "SECURITY.md"]
external/oasis-core/CHANGELOG.md:4807  [File missing] [Context: "../docs/oasis-node/metrics.md"]

lukaw3d commented:
I found a good approximation for detecting bad links. Find any `../../` in build folder:
`grep -o -E '\.\./\.\./.{0,50}' -r ./build; test $? -eq 1`

And even better, find all `href:"relative` in build folder:
`grep -o -E 'href:".{0,50}' -r ./build | grep -v -E 'href:"(https:|http:|mailto:|/|#)';  test $? -eq 1`
